<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Running the FastQC Pipeline with sbgr • sbgr</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha256-eZrrJcwDc/3uDhsdt61sL2oOBY362qM3lon1gyExkL0=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- sticky kit --><script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet">
<meta property="og:title" content="Running the FastQC Pipeline with sbgr">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">sbgr</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.1.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/sbgr.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/easy-api.html">Easy API: A User-Friendly Cascading API</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/nanxstats/sbgr">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>Running the FastQC Pipeline with sbgr</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/nanxstats/sbgr/blob/master/vignettes/sbgr.Rmd"><code>vignettes/sbgr.Rmd</code></a></small>
      <div class="hidden name"><code>sbgr.Rmd</code></div>

    </div>

    
    
<!-- override white background for classless code chunks -->
<style type="text/css" scoped>
  pre:not([class]) { background-color: #F8F8F8; }
  pre code { background-color: transparent; }
</style>
<div id="introduction" class="section level1">
<h1 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h1>
<p>This guide will get you started to use Seven Bridges API with the <em>R</em> client package <code>sbgr</code>, and guide you through the steps needed to run the <a href="https://igor.sbgenomics.com/public/pipelines/534520eed79f0049c0c9443a/">FastQC Analysis pipeline</a>, a simple public pipeline on the Seven Bridges Genomics platform.</p>
<p>The following primary steps will be included:</p>
<ul>
<li>uploading files and setting the metadata;</li>
<li>running the <a href="https://igor.sbgenomics.com/public/pipelines/534520eed79f0049c0c9443a/">FastQC</a> pipeline;</li>
<li>downloading the results.</li>
</ul>
</div>
<div id="preparatory-works" class="section level1">
<h1 class="hasAnchor">
<a href="#preparatory-works" class="anchor"></a>Preparatory works</h1>
<div id="install-the-package" class="section level2">
<h2 class="hasAnchor">
<a href="#install-the-package" class="anchor"></a>Install the package</h2>
<p>Install the latest development version from GitHub:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb1-1" title="1"><span class="co"># install.packages("devtools")  # if devtools was not installed</span></a>
<a class="sourceLine" id="cb1-2" title="2">devtools<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/devtools/topics/reexports">install_github</a></span>(<span class="st">"nanxstats/sbgr"</span>)</a></code></pre></div>
<p>The package runs under Microsoft Windows, OS X, and GNU/Linux. If you meet any problems when installing the package, please check the <a href="http://www.bioconductor.org/install/">Bioconductor - Install</a> page or <a href="https://github.com/nanxstats/sbgr/issues">create an issue on GitHub</a> and report the problem.</p>
<p>To load the package in <em>R</em>, simply use</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb2-1" title="1"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(<span class="st">"sbgr"</span>)</a></code></pre></div>
</div>
<div id="get-authentication-token" class="section level2">
<h2 class="hasAnchor">
<a href="#get-authentication-token" class="anchor"></a>Get authentication token</h2>
<p>The authentication token is used as an <em>password</em> to identify yourself when aceessing the API. Please remember to keep your authentication token secure.</p>
<p>To get your authentication token, please <a href="https://accounts.sbgenomics.com/">login</a> to the Seven Bridges Platform. Go to the <em>Developer - Auth Token</em> tab, then click the button to generate the auth token.</p>
<p>Alternatively, you could use the function <code><a href="../reference/misc_get_auth_token.html">misc_get_auth_token()</a></code> in <code>sbgr</code> to get the token and store it as an object in R workspace:</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb3-1" title="1">token =<span class="st"> </span><span class="kw"><a href="../reference/misc_get_auth_token.html">misc_get_auth_token</a></span>()</a></code></pre></div>
<pre><code>## Enter the generated authentication token:
## 1:</code></pre>
<p>This will automatically open the browser and redirect to the token generation page. You can copy and paste the token into the <em>R</em> console, press enter, and then the function will save the token as an character object <code>token</code> in the current workspace, which will make it convenient for you to use it in the next steps. For example:</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb5-1" title="1"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/print">print</a></span>(token)</a></code></pre></div>
<pre><code>## [1] "58aeb140197001306386001f5b34aa78"</code></pre>
<blockquote>
<p>The auth token can be continuously used as long as you do not manually regenerate it or disable it on the Seven Bridges platform.</p>
</blockquote>
</div>
</div>
<div id="create-projects" class="section level1">
<h1 class="hasAnchor">
<a href="#create-projects" class="anchor"></a>Create projects</h1>
<p>Pipelines need to be executed within projects. We can either use a project that has already been created, or we can use the API to create a new one. Let’s create one from scratch. We do this using the function <code><a href="../reference/project_new.html">project_new()</a></code>. The function requires a name for the project (we will use the name <code>'API tutorial'</code>) and a billing group, which could be retrived by the function <code><a href="../reference/billing.html">billing()</a></code>. You can also use the optional parameter <code>description</code> in <code><a href="../reference/project_new.html">project_new()</a></code> to set the project description.</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb7-1" title="1"><span class="co"># Get the first billing group that we have access to</span></a>
<a class="sourceLine" id="cb7-2" title="2"><span class="co"># and use it for the new project</span></a>
<a class="sourceLine" id="cb7-3" title="3">billing_group_id =<span class="st"> </span><span class="kw"><a href="../reference/billing.html">billing</a></span>(token)[[<span class="st">"items"</span>]][[<span class="dv">1</span>]][[<span class="st">"id"</span>]]</a>
<a class="sourceLine" id="cb7-4" title="4"></a>
<a class="sourceLine" id="cb7-5" title="5"><span class="co"># Create new project</span></a>
<a class="sourceLine" id="cb7-6" title="6">sbgr_project =<span class="st"> </span><span class="kw"><a href="../reference/project_new.html">project_new</a></span>(</a>
<a class="sourceLine" id="cb7-7" title="7">  token, <span class="dt">name =</span> <span class="st">"sbgr tutorial"</span>,</a>
<a class="sourceLine" id="cb7-8" title="8">  <span class="dt">description =</span> <span class="st">"sbgr tutorial project"</span>,</a>
<a class="sourceLine" id="cb7-9" title="9">  <span class="dt">billing_group_id =</span> billing_group_id)</a>
<a class="sourceLine" id="cb7-10" title="10"></a>
<a class="sourceLine" id="cb7-11" title="11"><span class="co"># The list sbgr_project now contains the ID of the new project,</span></a>
<a class="sourceLine" id="cb7-12" title="12"><span class="co"># which we will use in the subsequent calls</span></a>
<a class="sourceLine" id="cb7-13" title="13">sbgr_project_id =<span class="st"> </span>sbgr_project[[<span class="st">"id"</span>]]</a>
<a class="sourceLine" id="cb7-14" title="14"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/print">print</a></span>(sbgr_project_id)</a></code></pre></div>
<pre><code>## [1] "dbf274a0-6dc5-4453-be87-b16965add1aa"</code></pre>
<blockquote>
<p>The created project could be seen on the <a href="https://igor.sbgenomics.com/u/#q">Seven Bridges platform</a>.</p>
</blockquote>
</div>
<div id="get-project-details" class="section level1">
<h1 class="hasAnchor">
<a href="#get-project-details" class="anchor"></a>Get project details</h1>
<p>We will need the ID of a project to edit the file metatdata and execute pipelines. We can get the details of a project by using <code><a href="../reference/project_details.html">project_details()</a></code>.</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb9-1" title="1"><span class="kw"><a href="../reference/project_details.html">project_details</a></span>(token, <span class="dt">project_id =</span> sbgr_project_id)</a></code></pre></div>
<pre><code>$id
[1] "dbf274a0-6dc5-4453-be87-b16965add1aa"

$name
[1] "sbgr tutorial"

$description
[1] "sbgr tutorial project"

$my_permissions
$my_permissions$write
[1] TRUE

$my_permissions$copy
[1] TRUE

$my_permissions$execute
[1] TRUE

$my_permissions$admin
[1] TRUE</code></pre>
<p>The returned list contains the details of the project that you can access with your auth token. Each project is listed with its ID number, name, description, and your detailed permissions to access the project.</p>
</div>
<div id="upload-files-and-set-file-metadata" class="section level1">
<h1 class="hasAnchor">
<a href="#upload-files-and-set-file-metadata" class="anchor"></a>Upload files and set file metadata</h1>
<p>Download the archive of sample data files <a href="https://developer.sbgenomics.com/examples/download/sample1.tgz">sample1.tgz</a> and assume we have extracted the files to the current <em>R</em> working directory. We will upload the extracted file <code>sample1.fastq</code> in this example.</p>
<div id="upload-files-using-sbg-command-line-uploader" class="section level2">
<h2 class="hasAnchor">
<a href="#upload-files-using-sbg-command-line-uploader" class="anchor"></a>Upload files using SBG command line uploader</h2>
<p>The easiest way to upload files to the Seven Bridges Platform and set their metadata without using the GUI is via the <a href="https://developer.sbgenomics.com/tools/uploader/documentation">SBG command-line uploader</a>. Download the latest version of the <a href="https://igor.sbgenomics.com/sbg-uploader/sbg-uploader.tgz">SBG command line uploader</a>. If you have an older version of the uploader, please be sure to re-download and use the most recent version. You can also use the function <code><a href="../reference/misc_get_uploader.html">misc_get_uploader()</a></code> to download the uploader and extract it to a specified location.</p>
<p>Once we have prepared the uploader and the file, we can use the function <code><a href="../reference/misc_upload_cli.html">misc_upload_cli()</a></code> to call the CLI uploader for uploading files, by specifying the parameters in the function:</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb11-1" title="1"><span class="co"># download the SBG CLI uploader to home directory</span></a>
<a class="sourceLine" id="cb11-2" title="2"><span class="kw"><a href="../reference/misc_get_uploader.html">misc_get_uploader</a></span>(<span class="st">'~/'</span>)</a>
<a class="sourceLine" id="cb11-3" title="3"></a>
<a class="sourceLine" id="cb11-4" title="4"><span class="co"># suppose we have extracted sample1.fastq into current R working directory</span></a>
<a class="sourceLine" id="cb11-5" title="5">file_id =<span class="st"> </span><span class="kw"><a href="../reference/misc_upload_cli.html">misc_upload_cli</a></span>(</a>
<a class="sourceLine" id="cb11-6" title="6">  token, <span class="dt">project_id =</span> sbgr_project_id,</a>
<a class="sourceLine" id="cb11-7" title="7">  <span class="dt">file =</span> <span class="st">"sample1.fastq"</span>, <span class="dt">uploader =</span> <span class="st">'~/sbg-uploader/'</span>)</a></code></pre></div>
<p>After successful upload, the function will return the file’s ID number:</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb12-1" title="1"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/print">print</a></span>(file_id)</a></code></pre></div>
<pre><code>## [1] "559c41ebe4b0566fd159709a"</code></pre>
<blockquote>
<p>The uploaded file could be seen in the project we just created on the <a href="https://igor.sbgenomics.com/u/#q">Seven Bridges platform</a>.</p>
</blockquote>
<!--

## Using API Calls

Another way to upload files to the Seven Bridges Platform is via the API calls.

Uploading files with the Seven Bridges API is performed as a multi-step
operation. Basically, the required file is uploaded chunk by chunk,
and then assembled as a complete file (in a similar way as the AWS S3 API).
In order to upload a file, you will need to:

  1. Initialize the upload, defining the file name and size of the file.
  2. Get signed URLs for the upload of each of the file parts to Amazon
  Simple Storage Service (S3).
  3. Upload a part using a standard HTTP PUT call to the [Amazon S3pre-signed URL](http://docs.aws.amazon.com/AmazonS3/latest/dev/PresignedUrlUploadObject.html).
  4. If the part has been uploaded successfully, report completion of
  the part upload, otherwise handle the failure (retry, inform user, etc.).
  5. Repeat steps 2 to 5 for each file part.
  6. Once all file parts have been successfully uploaded, report completion
  of file upload.

In this example, we will upload only one file part, since our file is
a small one. For larger files, simply split them into smaller file parts
and repeat the steps shown below.

-->
<!-- TODO: Python tutorial works fine, httr has HTTP 403
     when doing the PUT() step. -->
<!--


```r
# initialize the upload of the file sample1.fastq
init = upload_init(
  token, project_id = sbgr_project_id,
  name = "sample1.fastq", size = file.size("sample1.fastq"))

upload_id = init[["upload_id"]]
partinfo = upload_info_part(token, upload_id = upload_id, part_number = 1)
uri = partinfo[["uri"]]
```

After we got the upload URL on Amazon S3, we could use the `httr` package to upload the file to it:


```r
library("httr")
fastqfile = upload_file(normalizePath("~/sample1.fastq"))
upload_s3 = PUT(uri, body = fastqfile)
# upload_s3 = requests.put(uri, data = f)
etag = upload_s3$headers[["ETag"]]
# We make the API call to confirm the successful upload
# to the S3 backing storage
upload_complete_part(token, upload_id, part_number = 1, e_tag = etag)
# We make the call to finalize the upload. This ensures that the
# file becomes available for other API requests.
upload_complete_all(token, upload_id)
```

-->
</div>
<div id="identify-uploaded-files" class="section level2">
<h2 class="hasAnchor">
<a href="#identify-uploaded-files" class="anchor"></a>Identify uploaded files</h2>
<p>To use the files you have uploaded in the previous step, we will first need to get their ID numbers. After uploading files to the project, we can obtain the file ID by running the <code><a href="../reference/file_list.html">file_list()</a></code> function to list all the files in the project at anytime.</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb14-1" title="1"><span class="kw"><a href="../reference/file_list.html">file_list</a></span>(token, <span class="dt">project_id =</span> sbgr_project_id)</a></code></pre></div>
<p>This will return a list about the details of the uploaded files in the project:</p>
<pre><code>## $items
## $items[[1]]
## $items[[1]]$id
## [1] "559c41ebe4b0566fd159709a"
##
## $items[[1]]$name
## [1] "sample1.fastq"
##
## $items[[1]]$size
## [1] 16
##
## $items[[1]]$origin
## $items[[1]]$origin$upload_name
## [1] "1436362037099"
##
##
## $items[[1]]$metadata
## $items[[1]]$metadata$file_type
## [1] "fastq"
##
## $items[[1]]$metadata$seq_tech
## [1] "illumina"
##
## $items[[1]]$metadata$sample
## [1] "sample1"</code></pre>
</div>
<div id="update-file-metadata" class="section level2">
<h2 class="hasAnchor">
<a href="#update-file-metadata" class="anchor"></a>Update file metadata</h2>
<p>Once the file is uploaded, we can use the function <code><a href="../reference/file_meta_update.html">file_meta_update()</a></code> to set or update the file metadata. After we got the file ID number, we can easily set the file metadata by specifying certain parameters in <code><a href="../reference/file_meta_update.html">file_meta_update()</a></code>. For example:</p>
<div class="sourceCode" id="cb16"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb16-1" title="1"><span class="kw"><a href="../reference/file_meta_update.html">file_meta_update</a></span>(</a>
<a class="sourceLine" id="cb16-2" title="2">  token, <span class="dt">project_id =</span> sbgr_project_id,</a>
<a class="sourceLine" id="cb16-3" title="3">  <span class="dt">file_id    =</span> file_id,</a>
<a class="sourceLine" id="cb16-4" title="4">  <span class="dt">file_type  =</span> <span class="st">"fastq"</span>, <span class="dt">seq_tech =</span> <span class="st">"Illumina"</span>,</a>
<a class="sourceLine" id="cb16-5" title="5">  <span class="dt">qual_scale =</span> <span class="st">"illumina18"</span>,</a>
<a class="sourceLine" id="cb16-6" title="6">  <span class="dt">sample     =</span> <span class="st">"example_human_Illumina"</span>,</a>
<a class="sourceLine" id="cb16-7" title="7">  <span class="dt">library    =</span> <span class="st">"Test"</span>, <span class="dt">paired_end =</span> <span class="st">"1"</span>)</a></code></pre></div>
<pre><code>## $id
## [1] "559c41ebe4b0566fd159709a"
##
## $name
## [1] "sample1.fastq"
##
## $size
## [1] 16
##
## $origin
## $origin$upload_name
## [1] "1436361569553"
##
##
## $metadata
## $metadata$file_type
## [1] "fastq"
##
## $metadata$qual_scale
## [1] "illumina18"
##
## $metadata$seq_tech
## [1] "Illumina"
##
## $metadata$sample
## [1] "example_human_Illumina"
##
## $metadata$library
## [1] "Test"
##
## $metadata$paired_end
## [1] "1"</code></pre>
<p>The new metadata of the file will be shown after being succesfully updated.</p>
<p>The <code>sbgr</code> package also has a helper function <code><a href="../reference/misc_make_metadata.html">misc_make_metadata()</a></code> to generate metadata files. For more information about metadata, please refer to the <a href="https://developer.sbgenomics.com/platform/metadata">file metadata documentation</a>.</p>
</div>
</div>
<div id="list-public-pipelines-and-add-pipelines-to-project" class="section level1">
<h1 class="hasAnchor">
<a href="#list-public-pipelines-and-add-pipelines-to-project" class="anchor"></a>List public pipelines and add pipelines to project</h1>
<p>We want to run the <a href="https://igor.sbgenomics.com/public/pipelines/534520eed79f0049c0c9443a/">FastQC pipeline</a> in our project, so we will need its ID number to identify it. Since the FastQC pipeline is available in the <a href="https://igor.sbgenomics.com/public/pipelines">Seven Bridges public pipelines repository</a>, we will now list all of these pipelines along with their IDs to find the ID that we need. We can list all public pipelines via <code><a href="../reference/pipeline_list_pub.html">pipeline_list_pub()</a></code>:</p>
<div class="sourceCode" id="cb18"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb18-1" title="1">pipeline_list =<span class="st"> </span><span class="kw"><a href="../reference/pipeline_list_pub.html">pipeline_list_pub</a></span>(token)</a>
<a class="sourceLine" id="cb18-2" title="2"><span class="co"># convert public pipeline list to a matrix</span></a>
<a class="sourceLine" id="cb18-3" title="3">pipeline_mat =<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/matrix">matrix</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/unlist">unlist</a></span>(pipeline_list<span class="op">$</span>items), <span class="dt">ncol =</span> <span class="dv">2</span>, <span class="dt">byrow =</span> <span class="ot">TRUE</span>)</a>
<a class="sourceLine" id="cb18-4" title="4"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/print">print</a></span>(pipeline_mat)</a></code></pre></div>
<pre><code>## [,1]                       [,2]
## [1,] "534522d3d79f0049c0c9444d" "Targeted Capture Analysis - BWA + GATK 2.3.9-Lite (with Metrics)"
## [2,] "534522f6d79f0049c0c9444e" "Whole Exome Analysis - BWA + GATK 2.3.9-Lite (with Metrics)"
## [3,] "5345230dd79f0049c0c9444f" "Whole Genome Analysis - BWA + GATK 2.3.9-Lite (with Metrics)"
## [4,] "534521e5d79f0049c0c94445" "RNA-Seq Alignment for Ion Proton - TopHat + Bowtie 2"
## [5,] "53452154d79f0049c0c94442" "RNA-Seq Alignment - TopHat"
## [6,] "546ba1bbd79f00701cb276a6" "Cellular Research Precise Analysis Pipeline"
## [7,] "5392689dd79f007259672c79" "Amplicon Experiment QC"
## [8,] "53927c20d79f007259672c8e" "Amplicon Experiment QC - AmpliSeq Exomes"
## [9,] "534521d1d79f0049c0c94444" "RNA-Seq Alignment for Ion Proton - STAR + Bowtie 2"
## [10,] "53452130d79f0049c0c94441" "RNA-Seq Alignment - STAR"
## [11,] "53452273d79f0049c0c94448" "RNA-Seq Differential Expression - Cuffdiff (with Visualization)"
## [12,] "540dd19dd79f00766c174ead" "Fusion Transcript Detection - ChimeraScan"
## [13,] "5346ba3dd79f0049c0c944a6" "RNA-Seq De Novo Assembly - Trinity"
## [14,] "540dd2fad79f00766c174eb0" "Fusion Transcript Detection - STAR + Chimera"
## [15,] "534521f5d79f0049c0c94446" "RNA-Seq De Novo Assembly and Analysis - Trinity"
## [16,] "534520eed79f0049c0c9443a" "FastQC Analysis"
## [17,] "5345222cd79f0049c0c94447" "RNA-Seq Differential Expression - Cuffdiff"
## [18,] "53b2e456d79f004c55605245" "Trio analysis pipeline (Whole Exome)"
## [19,] "53b2d73bd79f004c5560523d" "Trio analysis pipeline (Whole Genome)"
## [20,] "53b2a2f0d79f004c55605227" "Illumina BAM to FASTQ"
## [21,] "534522b0d79f0049c0c9444b" "Stanford HugeSeq WGS - from Aligned Reads (BAM)"
## [22,] "53452dd2d79f0049c0c94459" "Alignment Metrics QC"
## [23,] "534522c4d79f0049c0c9444c" "Stanford HugeSeq WGS - from Unaligned Reads (FASTQ)"
## [24,] "53452299d79f0049c0c9444a" "Stanford HugeSeq WGS - Structural Variation Only"
## [25,] "53452289d79f0049c0c94449" "Stanford HugeSeq WGS - Skip Alignment"
## [26,] "5345210fd79f0049c0c94440" "On-target variant selection and QC"
## [27,] "534520ffd79f0049c0c9443b" "Merge FASTQ Files"
## [28,] "534520dad79f0049c0c94439" "Exome Variant Quality Score Recalibration - GATK 2.3.9-Lite"
## [29,] "534520c9d79f0049c0c94438" "Exome SNP calling Ion Torrent - BWA - MEM + GATK 2.3.9-Lite (with Metrics)"
## [30,] "534520b5d79f0049c0c94437" "Exome SNP Calling - Mosaik + FreeBayes"
## [31,] "534520a5d79f0049c0c94432" "Exome Coverage QC"
## [32,] "53452096d79f0049c0c94431" "Exome Analysis - BWA + GATK 1.6"
## [33,] "53452077d79f0049c0c94430" "Convert SFF files to FASTQ files"
## [34,] "5345205cd79f0049c0c9442f" "Convert SAM/BAM to FASTQ"</code></pre>
<p>The pipeline we want is “FastQC Analysis”, so we will copy it to our project using the function <code><a href="../reference/pipeline_add.html">pipeline_add()</a></code>:</p>
<div class="sourceCode" id="cb20"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb20-1" title="1"><span class="co"># Get the FastQC pipeline ID number in pipeline_mat</span></a>
<a class="sourceLine" id="cb20-2" title="2"><span class="kw"><a href="../reference/pipeline_add.html">pipeline_add</a></span>(</a>
<a class="sourceLine" id="cb20-3" title="3">  token, <span class="dt">project_id_to =</span> sbgr_project_id,</a>
<a class="sourceLine" id="cb20-4" title="4">  <span class="dt">pipeline_id =</span> pipeline_mat[<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/which">which</a></span>(pipeline_mat[, <span class="dv">2</span>] <span class="op">==</span><span class="st"> "FastQC Analysis"</span>), <span class="dv">1</span>])</a></code></pre></div>
<p>This call returns the details of the copied pipeline.</p>
<pre><code>## $id
## [1] "559c3848896a5d236e19d4a2"
##
## $name
## [1] "FastQC Analysis"
##
## $description
## [1] "The FastQC tool, developed by the Babraham Institute, analyzes sequence data from FASTQ, BAM, or SAM files. It produces a set of metrics and charts that help identify technical problems with the data. It's a good idea to run this pipeline on files you receive from a sequencer or a collaborator to get a general idea of how well the sequencing experiment went. Results from this pipeline can inform if and how you should proceed with your analysis."
##
## $revision
## [1] "0"</code></pre>
<p>The added pipeline could also be seen in the project we created on the <a href="https://igor.sbgenomics.com/u/">Seven Bridges platform</a>.</p>
<blockquote>
<p>Note that the pipeline now has a <strong>new</strong> ID number. This identifies the pipeline within your project. The ID of a pipeline when listed in the public repository is <strong>different</strong> from its ID when listed inside your project.</p>
</blockquote>
</div>
<div id="get-pipeline-details-and-find-input-nodes" class="section level1">
<h1 class="hasAnchor">
<a href="#get-pipeline-details-and-find-input-nodes" class="anchor"></a>Get pipeline details and find input nodes</h1>
<p>Assume we have cloned the <a href="https://igor.sbgenomics.com/public/pipelines/534520eed79f0049c0c9443a/">FastQC pipeline</a> into our project via the above code or via the SBG platform. Now we need the <strong>new</strong> ID number of the FastQC pipeline in our project, so we can execute it via the API. To get the details of all available pipelines in your project, use <code><a href="../reference/pipeline_list_project.html">pipeline_list_project()</a></code>:</p>
<div class="sourceCode" id="cb22"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb22-1" title="1">project_pipeline =<span class="st"> </span><span class="kw"><a href="../reference/pipeline_list_project.html">pipeline_list_project</a></span>(token, <span class="dt">project_id =</span> sbgr_project_id)</a>
<a class="sourceLine" id="cb22-2" title="2"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/print">print</a></span>(project_pipeline)</a></code></pre></div>
<p>This will return a list of the pipelines in the specified project:</p>
<pre><code>## $items
## $items[[1]]
## $items[[1]]$id
## [1] "559c3848896a5d236e19d4a2"
##
## $items[[1]]$name
## [1] "FastQC Analysis"</code></pre>
<p>Now we can extract the ID number of the pipeline:</p>
<div class="sourceCode" id="cb24"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb24-1" title="1">pipeline_id =<span class="st"> </span>project_pipeline[[<span class="st">"items"</span>]][[<span class="dv">1</span>]][[<span class="st">"id"</span>]]</a></code></pre></div>
<p>To input the file we uploaded to a pipeline, we have to get some more of the pipeline’s details. In particular, we will need to get the ID number of its <em>input nodes</em>. We can list the full details of our chosen pipeline inside the project using the function <code><a href="../reference/pipeline_details.html">pipeline_details()</a></code>:</p>
<div class="sourceCode" id="cb25"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb25-1" title="1">fastqc_details =<span class="st"> </span><span class="kw"><a href="../reference/pipeline_details.html">pipeline_details</a></span>(</a>
<a class="sourceLine" id="cb25-2" title="2">  token, <span class="dt">project_id =</span> sbgr_project_id, <span class="dt">pipeline_id =</span> pipeline_id)</a>
<a class="sourceLine" id="cb25-3" title="3"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/print">print</a></span>(fastqc_details)</a></code></pre></div>
<p>The returned list contains the information of the pipeline. We can see that our chosen pipeline contains a single input node with ID <code>177252</code>:</p>
<pre><code>## $id
## [1] "559c3848896a5d236e19d4a2"
##
## $name
## [1] "FastQC Analysis"
##
## $description
## [1] "The FastQC tool, developed by the Babraham Institute, analyzes sequence data from FASTQ, BAM, or SAM files. It produces a set of metrics and charts that help identify technical problems with the data. It's a good idea to run this pipeline on files you receive from a sequencer or a collaborator to get a general idea of how well the sequencing experiment went. Results from this pipeline can inform if and how you should proceed with your analysis."
##
## $revision
## [1] "0"
##
## $inputs
## $inputs[[1]]
## $inputs[[1]]$id
## [1] 177252
##
## $inputs[[1]]$name
## [1] "FASTQ Reads"
##
## $inputs[[1]]$required
## [1] TRUE
##
## $inputs[[1]]$accepts_list
## [1] FALSE
##
## $inputs[[1]]$file_types
## $inputs[[1]]$file_types[[1]]
## [1] "bam"
##
## $inputs[[1]]$file_types[[2]]
## [1] "fastq"
##
## $inputs[[1]]$file_types[[3]]
## [1] "sam"
##
##
##
##
## $nodes
## $nodes[[1]]
## $nodes[[1]]$id
## [1] 680142
##
## $nodes[[1]]$name
## [1] "FastQC"
##
## $nodes[[1]]$parameters
## list()
##
##
##
## $outputs
## $outputs[[1]]
## $outputs[[1]]$id
## [1] 556132
##
## $outputs[[1]]$name
## [1] "FastQC Reports Archive"
##
##
## $outputs[[2]]
## $outputs[[2]]$id
## [1] 1274412
##
## $outputs[[2]]$name
## [1] "FastQC Charts"</code></pre>
<p>This call returns a list of information about the FastQC pipeline. In particular, it gives the following information:</p>
<ul>
<li>
<code>id</code>: The ID number of the pipeline, used by the API to uniquely refer to it</li>
<li>
<code>revision</code>: The revision of the pipeline</li>
<li>
<code>name</code>: The human-readable name of the pipeline</li>
<li>
<code>description</code>: A short description of what the pipeline’s function</li>
<li>
<code>inputs</code>: An array listing the pipeline’s input nodes, i.e. those pipeline apps which require files as input</li>
<li>
<code>nodes</code>: An array of intermediary nodes, representing the pipeline apps which transform the data. These may have editable parameters that should be set before running the pipeline, although in this particular pipeline example no parameters are editable.</li>
<li>
<code>output</code>: An array of output nodes, representing apps that produce files, reports, visualizations, etc.</li>
</ul>
<p>We will need to pay special attention to the sub-string that is a list of key-value pairs marked <code>inputs</code> for pipeline input nodes. Under each <code>input</code> node is a list of suggested files. These are the publicly available files that Seven Bridges Genomics recommends providing as input. Typically these are reference files, such as <em>reference genomes</em>, <em>SNP databases</em>, <em>known indels</em>, etc.</p>
<p>By inspecting the <code>inputs</code> part of the response to the call we just made, we can see the names and ID numbers of the suggested files for each input app. Notice that there is an app named <code>FASTQ</code>, which takes fastq files as input. This will be where we input the file <code>sample1.fastq</code> we want to analyze. We could extract and store the input id as <code>input_id</code>:</p>
<div class="sourceCode" id="cb27"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb27-1" title="1">input_id =<span class="st"> </span>fastqc_details<span class="op">$</span>inputs[[<span class="dv">1</span>]]<span class="op">$</span>id</a></code></pre></div>
</div>
<div id="run-the-pipeline" class="section level1">
<h1 class="hasAnchor">
<a href="#run-the-pipeline" class="anchor"></a>Run the pipeline</h1>
<p>Now we have collected all the necessary data to run the pipeline:</p>
<table class="table">
<thead><tr class="header">
<th>Data</th>
<th>R Object</th>
<th>ID Number</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>Project ID</td>
<td><code>sbgr_project_id</code></td>
<td><code>dbf274a0-6dc5-4453-be87-b16965add1aa</code></td>
</tr>
<tr class="even">
<td>Pipeline ID</td>
<td><code>pipeline_id</code></td>
<td><code>559c3848896a5d236e19d4a2</code></td>
</tr>
<tr class="odd">
<td>Input ID</td>
<td><code>input_id</code></td>
<td><code>177252</code></td>
</tr>
<tr class="even">
<td>File ID</td>
<td><code>file_id</code></td>
<td><code>559c41ebe4b0566fd159709a</code></td>
</tr>
</tbody>
</table>
<p>We can then run the pipeline task using <code><a href="../reference/task_run.html">task_run()</a></code> to execute a task:</p>
<div class="sourceCode" id="cb28"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb28-1" title="1"><span class="co"># put input_id and file_id into list</span></a>
<a class="sourceLine" id="cb28-2" title="2">inputs =<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/list">list</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/list">list</a></span>(file_id))</a>
<a class="sourceLine" id="cb28-3" title="3"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/names">names</a></span>(inputs) =<span class="st"> </span>input_id</a>
<a class="sourceLine" id="cb28-4" title="4"></a>
<a class="sourceLine" id="cb28-5" title="5"><span class="co"># put the task info into a list</span></a>
<a class="sourceLine" id="cb28-6" title="6">task =<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/list">list</a></span>(</a>
<a class="sourceLine" id="cb28-7" title="7">  <span class="st">"name"</span> =<span class="st"> "FastQC with sbgr"</span>,</a>
<a class="sourceLine" id="cb28-8" title="8">  <span class="st">"description"</span> =<span class="st"> "FastQC task with sbgr"</span>,</a>
<a class="sourceLine" id="cb28-9" title="9">  <span class="st">"pipeline_id"</span> =<span class="st"> </span>pipeline_id,</a>
<a class="sourceLine" id="cb28-10" title="10">  <span class="st">"inputs"</span> =<span class="st"> </span>inputs)</a>
<a class="sourceLine" id="cb28-11" title="11"></a>
<a class="sourceLine" id="cb28-12" title="12"><span class="co"># add and run the task</span></a>
<a class="sourceLine" id="cb28-13" title="13">fastqc_task =<span class="st"> </span><span class="kw"><a href="../reference/task_run.html">task_run</a></span>(token, <span class="dt">project_id =</span> sbgr_project_id, <span class="dt">task_details =</span> task)</a>
<a class="sourceLine" id="cb28-14" title="14"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/print">print</a></span>(fastqc_task)</a></code></pre></div>
<p>If the request is well-constructed and contains valid IDs, it will return details of the newly-created task as a response:</p>
<pre><code>## $id
## [1] "2bfb18cc-f969-43a1-af75-8d63c1e0f54a"
##
## $name
## [1] "FastQC with sbgr"
##
## $description
## [1] "FastQC task with sbgr"
##
## $pipeline_id
## [1] "559c3848896a5d236e19d4a2"
##
## $pipeline_revision
## [1] "0"
##
## $start_time
## [1] 1.436304e+12
##
## $status
## $status$status
## [1] "active"
##
##
## $inputs
## $inputs$`177252`
## $inputs$`177252`[[1]]
## [1] ":559c41ebe4b0566fd159709a"
##
##
##
## $parameters
## $parameters$`680142`
## $parameters$`680142`$files
## NULL
##
## $parameters$`680142`$value
## named list()
##
##
##
## $outputs
## $outputs$`556132`
## list()
##
## $outputs$`1274412`
## list()</code></pre>
</div>
<div id="monitor-the-task-and-download-the-results" class="section level1">
<h1 class="hasAnchor">
<a href="#monitor-the-task-and-download-the-results" class="anchor"></a>Monitor the task and download the results</h1>
<p>Tasks typically take a long time to finish. We can periodically (say, every 30 seconds) check the status of a task by using the function <code><a href="../reference/task_details.html">task_details()</a></code>:</p>
<div class="sourceCode" id="cb30"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb30-1" title="1"><span class="cf">while</span> (<span class="ot">TRUE</span>) {</a>
<a class="sourceLine" id="cb30-2" title="2">  task_running =<span class="st"> </span><span class="kw"><a href="../reference/task_details.html">task_details</a></span>(token, <span class="dt">project_id =</span> sbgr_project_id,</a>
<a class="sourceLine" id="cb30-3" title="3">                              <span class="dt">task_id =</span> fastqc_task<span class="op">$</span>id)</a>
<a class="sourceLine" id="cb30-4" title="4">  <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/cat">cat</a></span>(<span class="st">"Running FastQC task...</span><span class="ch">\n</span><span class="st">"</span>)</a>
<a class="sourceLine" id="cb30-5" title="5">  <span class="cf">if</span> (task_running[[<span class="st">"status"</span>]][[<span class="st">"status"</span>]] <span class="op">!=</span><span class="st"> "active"</span>) <span class="cf">break</span></a>
<a class="sourceLine" id="cb30-6" title="6">  <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/Sys.sleep">Sys.sleep</a></span>(<span class="dv">30</span>)</a>
<a class="sourceLine" id="cb30-7" title="7">}</a>
<a class="sourceLine" id="cb30-8" title="8"></a>
<a class="sourceLine" id="cb30-9" title="9"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/print">print</a></span>(task_running)</a></code></pre></div>
<p>Usually after running for several minutes, the task should be finished, and you will receive emails when the task is finished. Alternatively, you could also monitor the task status in your dashboard on the SBG website. The output would contain the task information:</p>
<pre><code>## $id
## [1] "4bf0a677-cc90-4eb8-bcb7-5759adc36449"
##
## $name
## [1] "FastQC with sbgr"
##
## $description
## [1] "FastQC task with sbgr"
##
## $pipeline_id
## [1] "559c3848896a5d236e19d4a2"
##
## $pipeline_revision
## [1] "0"
##
## $start_time
## [1] 1.43629e+12
##
## $status
## $status$status
## [1] "completed"
##
## $status$message
## [1] "Completed."
##
##
## $inputs
## $inputs$`177252`
## $inputs$`177252`[[1]]
## [1] ":559c41ebe4b0566fd159709a"
##
##
##
## $parameters
## $parameters$`680142`
## $parameters$`680142`$files
## NULL
##
## $parameters$`680142`$value
## named list()
##
##
##
## $outputs
## $outputs$`556132`
## $outputs$`556132`[[1]]
## [1] "559c44dbe4b07462e814cb7c"
##
##
## $outputs$`1274412`
## $outputs$`1274412`[[1]]
## [1] "559c44dbe4b0566fd1597562"</code></pre>
<p>Once a task has finished, you can locate and download its output files. You will need an external program like <code>wget</code> to perform the download. We use the function <code><a href="../reference/file_download_url.html">file_download_url()</a></code> to return the download URLs of the output file (for example, the first output file):</p>
<div class="sourceCode" id="cb32"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb32-1" title="1">report_url =<span class="st"> </span><span class="kw"><a href="../reference/file_download_url.html">file_download_url</a></span>(</a>
<a class="sourceLine" id="cb32-2" title="2">  token, <span class="dt">project_id =</span> sbgr_project_id,</a>
<a class="sourceLine" id="cb32-3" title="3">  <span class="dt">file_id =</span> task_running[[<span class="st">"outputs"</span>]][[<span class="dv">1</span>]][[<span class="dv">1</span>]])</a>
<a class="sourceLine" id="cb32-4" title="4"><span class="co"># download the output archive using wget</span></a>
<a class="sourceLine" id="cb32-5" title="5"><span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/download.file">download.file</a></span>(</a>
<a class="sourceLine" id="cb32-6" title="6">  <span class="dt">url =</span> report_url[[<span class="st">"url"</span>]], <span class="dt">destfile =</span> <span class="st">"fastqc_report.zip"</span>, <span class="dt">method =</span> <span class="st">"wget"</span>)</a>
<a class="sourceLine" id="cb32-7" title="7"><span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/untar">untar</a></span>(<span class="st">"fastqc_report.zip"</span>)  <span class="co"># extract the zipped files</span></a>
<a class="sourceLine" id="cb32-8" title="8"><span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/browseURL">browseURL</a></span>(<span class="st">"sample1_fastqc/fastqc_report.html"</span>)  <span class="co"># open html report in browser</span></a></code></pre></div>
</div>
<div id="additional-resources" class="section level1">
<h1 class="hasAnchor">
<a href="#additional-resources" class="anchor"></a>Additional resources</h1>
<p>The <a href="https://developer.sbgenomics.com">SBG Developer Hub</a> provides API Documentation, quickstart guide, SDK tutorials and additional information on developing your own applications using the Seven Bridges platform.</p>
<p>For a related example of running a series of tasks using a fixed pipeline and parameters, download the archive <a href="https://developer.sbgenomics.com/examples/download/sbg_api_example.tar.gz">sbg_api_example.tar.gz</a> which contains sample files and a full worked example in Python.</p>
</div>
<div id="acknowledgement" class="section level1">
<h1 class="hasAnchor">
<a href="#acknowledgement" class="anchor"></a>Acknowledgement</h1>
<p>This documentation is based on the original Python version of the API documentation and tutorials. We gratefully appreciate the support and excellent job from the team at Seven Bridges Genomics, Inc.</p>
<hr>
<p>© Seven Bridges Genomics, Inc. 2019</p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#introduction">Introduction</a></li>
      <li>
<a href="#preparatory-works">Preparatory works</a><ul class="nav nav-pills nav-stacked">
<li><a href="#install-the-package">Install the package</a></li>
      <li><a href="#get-authentication-token">Get authentication token</a></li>
      </ul>
</li>
      <li><a href="#create-projects">Create projects</a></li>
      <li><a href="#get-project-details">Get project details</a></li>
      <li>
<a href="#upload-files-and-set-file-metadata">Upload files and set file metadata</a><ul class="nav nav-pills nav-stacked">
<li><a href="#upload-files-using-sbg-command-line-uploader">Upload files using SBG command line uploader</a></li>
      <li><a href="#identify-uploaded-files">Identify uploaded files</a></li>
      <li><a href="#update-file-metadata">Update file metadata</a></li>
      </ul>
</li>
      <li><a href="#list-public-pipelines-and-add-pipelines-to-project">List public pipelines and add pipelines to project</a></li>
      <li><a href="#get-pipeline-details-and-find-input-nodes">Get pipeline details and find input nodes</a></li>
      <li><a href="#run-the-pipeline">Run the pipeline</a></li>
      <li><a href="#monitor-the-task-and-download-the-results">Monitor the task and download the results</a></li>
      <li><a href="#additional-resources">Additional resources</a></li>
      <li><a href="#acknowledgement">Acknowledgement</a></li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by Nan Xiao, Tengfei Yin, Seven Bridges Genomics.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.0.</p>
</div>
      </footer>
</div>

  

  </body>
</html>
